src/main/java/com/google/sps/servlets/GoServlet.java

```diff
@@ -0,0 +1,316 @@
+        minTravelTime = travelTime;
+        mostOptimalWaypointCombination = waypointCombination;
+      }
+    }
+    return mostOptimalWaypointCombination;
+  }
+
+  /**
+   * Finds the most optimal route of travel between the origin and destination and a set of
+   * waypoints which if the exact location is not known, an exact location is determined and chosen
+   * based on the resulting travel time. Scope of method is public for testing purposes.
+   *
+   * @param origin The starting point of travel.
+   * @param destination The ending point of travel.
+   * @param waypoints The waypoints that should be visited while travelling from the origin to the
+   *     destination.
+   * @return The most optimal set of waypoints between origin and destination.
+   * @throws GeocodingException An exception thrown when an error occurs with the Geocoding API.
+   * @throws PlacesException An exception thrown when an error occurs with the Places API.
+   * @throws DirectionsException An exception thrown when an error occurs with the Directions API.
+   */
+  public List<String> optimizeSearchNearbyWaypoints(
+      String origin, String destination, List<String> waypoints)
+      throws GeocodingException, PlacesException, DirectionsException {
+
+    Optional<LatLng> originAsCoordinates =
+        GeocodingResultUtility.getCoordinates(
+            geocodingClientFactory.getGeocodingClient(apiKey).getGeocodingResult(origin));
+
+    Optional<LatLng> destinationAsCoordinates =
+        GeocodingResultUtility.getCoordinates(
+            geocodingClientFactory.getGeocodingClient(apiKey).getGeocodingResult(destination));
+
+    if (!originAsCoordinates.isPresent() || !destinationAsCoordinates.isPresent()) {
+      throw new GeocodingException("Origin or destination is invalid");
+    }
+
+    List<String> streetAddressWaypoints = new ArrayList<>();
+    List<Optional<LatLng>> streetAddressWaypointsAsCoordinates = new ArrayList<>();
+    List<Optional<PlaceType>> nonStreetAddressWaypointsAsPlaceTypes = new ArrayList<>();
+    separateWaypoints(
+        waypoints,
+        streetAddressWaypoints,
+        streetAddressWaypointsAsCoordinates,
+        nonStreetAddressWaypointsAsPlaceTypes);
+
+    // All street address coordinates including origin and destination are collected
+    List<Optional<LatLng>> streetAddressesAsCoordinates = new ArrayList<>();
+    streetAddressesAsCoordinates.addAll(streetAddressWaypointsAsCoordinates);
+    streetAddressesAsCoordinates.add(originAsCoordinates);
+    streetAddressesAsCoordinates.add(destinationAsCoordinates);
+
+    // Remove all optional empty entries of street address coordinates and non street address
+    // waypoints
+    List<LatLng> nonEmptyStreetAddressesAsCoordinates =
+        streetAddressesAsCoordinates.stream()
+            .filter(coordinates -> !coordinates.equals(Optional.empty()))
+            .map(coordinates -> coordinates.get())
+            .collect(Collectors.toList());
```

Comment:
maybe create an auxiliary method to do that? It could accept `List<Optional<T>>` so you can use it for both lists

Reply:
Gonna ask you on chat - not entirely sure how to set types in Java templates

Reply:
Fixed, thanks! Like you mentioned, was just missing a <T>

Reply:
`RESOLVED`